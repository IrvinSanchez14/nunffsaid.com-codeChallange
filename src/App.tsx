import { Snackbar } from '@material-ui/core';
import React, { useCallback, useState } from 'react';
import { useEffect } from 'react';
import styled from 'styled-components';
import generateMessage, { Message } from './Api';
import Board from './Board';
import ButtonStyle from './ButtonStyle';

const App: React.FC<{}> = () => {
  const [messages, setMessages] = useState<Message[]>([]);
  const [runningMessage, setRunningMessage] = useState(true)
  const [open, setOpen] = useState(false)
  const [errorMessage, setErrorMessage] = useState('')
  const [errorIndex, setErrorIndex] = useState<number[]>([])

  useEffect(() => {
    if (runningMessage) {
      const cleanUp = generateMessage((message: Message) => {
        setMessages(oldMessages => [...oldMessages, message]);
      });
      return cleanUp;
    }
  }, [setMessages, runningMessage]);

  const stopMessage = useCallback(() => {
    setRunningMessage(!runningMessage)
  },[setRunningMessage, runningMessage])

  const clearMessages = () => {
    setMessages([])
  }

  const handleClose = (event?: React.SyntheticEvent | Event, reason?: string) => {
    if (reason === 'clickaway') {
      return;
    }

    setOpen(false);
  };

  const showNotification = useCallback(() => {
    const setMessageError = messages.map((value, index) => {
      return {
        ...value,
        index
      }
    }).sort((a,b) => b.index - a.index).find((item) => item.priority === 0 && !errorIndex.includes(item.index))
    if (setMessageError) {
      setOpen(true)
      setErrorIndex(oldIndex => [...oldIndex, setMessageError.index])
      setErrorMessage(setMessageError.message)
    }
  },[messages, setOpen, setErrorMessage, setErrorIndex, errorIndex])

  useEffect(() => {
    showNotification()
  }, [showNotification])

  return (
    <div>
      <StyledTitle>nunffsaid.com Coding Challenge</StyledTitle>
      <Snackbar open={open} autoHideDuration={2000} anchorOrigin={{ vertical: 'top', horizontal: 'right'}} onClose={handleClose}>
        <StyledContentSnack>
          {errorMessage}
          <StyledCloseSnack onClick={handleClose}>X</StyledCloseSnack>
        </StyledContentSnack>
      </Snackbar>
      <StyledContainerButton>
        <ButtonStyle label={runningMessage ? 'STOP' : 'START'} onClick={stopMessage} />
        <ButtonStyle label="CLEAR" onClick={clearMessages} />
      </StyledContainerButton>
        <Board updatedMessage={setMessages} information={messages} />
      <StyledContainerBoard>
      </StyledContainerBoard>
    </div>
  );
}

const StyledTitle = styled.h1`
  color: #000;
`

const StyledContainerButton = styled.div`
  display: flex;
  justify-content: center;
`

const StyledContainerBoard = styled.div`
  margin: 50px 0px 0px 0px;
`

const StyledContentSnack = styled.div`
  border-radius: 4px;
  box-shadow: rgb(0 0 0 / 20%) 0px 3px 5px -1px, rgb(0 0 0 / 14%) 0px 6px 10px 0px, rgb(0 0 0 / 12%) 0px 1px 18px 0px;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  background-color: #f56236;
  display: flex;
  padding: 6px 16px;
`

const StyledCloseSnack = styled.span`
  margin-left: 15px;
  cursor: pointer;
  font-weight: bold;
`

export default App;
