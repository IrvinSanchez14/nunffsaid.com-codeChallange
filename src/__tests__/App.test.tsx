import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import App from '../App';

test('renders learn react link', () => {
  const comp = render(<App />);
  expect(comp).toBeTruthy();
});

describe('App functionality', () => {
  it('should render the components correctly', () => {
    const comp = render(<App />);
    expect(comp).toMatchSnapshot();
  })

  it('should change the value of button to START', () => {
    const { getByText } = render(<App />);
    const buttonStop = getByText("STOP")
    fireEvent.click(buttonStop)

    expect(getByText("START")).toBeTruthy();
  })
})
