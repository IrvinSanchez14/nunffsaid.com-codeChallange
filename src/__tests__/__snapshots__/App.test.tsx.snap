// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App functionality should render the components correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1
          class="sc-iBPTik kXpFsL"
        >
          nunffsaid.com Coding Challenge
        </h1>
        <div
          class="sc-fubCzh ibgszL"
        >
          <div
            class="sc-jSgvzq cSyPzI"
          >
            <button
              class="sc-gKseQn dquRBr"
            >
              STOP
            </button>
          </div>
          <div
            class="sc-jSgvzq cSyPzI"
          >
            <button
              class="sc-gKseQn dquRBr"
            >
              CLEAR
            </button>
          </div>
        </div>
        <div
          class="sc-hKgJUU bXmkWD"
        >
          <div
            class="sc-eCstlR hYyfgm"
          >
            <div>
              Error Type 1
            </div>
            <div>
              Count 0
            </div>
          </div>
          <div
            class="sc-eCstlR hYyfgm"
          >
            <div>
              Warning Type 2
            </div>
            <div>
              Count 0
            </div>
          </div>
          <div
            class="sc-eCstlR hYyfgm"
          >
            <div>
              Info Type 3
            </div>
            <div>
              Count 1
            </div>
            <div
              class="sc-bdfBQB kdxtmC"
              type="2"
            >
              <span
                class="sc-gsTEea hKnFKv"
              >
                Provident omnis voluptas modi vel enim.
              </span>
              <span
                class="sc-dlfnuX einvaF"
              >
                Clear
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-pGacB fTQQLE"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1
        class="sc-iBPTik kXpFsL"
      >
        nunffsaid.com Coding Challenge
      </h1>
      <div
        class="sc-fubCzh ibgszL"
      >
        <div
          class="sc-jSgvzq cSyPzI"
        >
          <button
            class="sc-gKseQn dquRBr"
          >
            STOP
          </button>
        </div>
        <div
          class="sc-jSgvzq cSyPzI"
        >
          <button
            class="sc-gKseQn dquRBr"
          >
            CLEAR
          </button>
        </div>
      </div>
      <div
        class="sc-hKgJUU bXmkWD"
      >
        <div
          class="sc-eCstlR hYyfgm"
        >
          <div>
            Error Type 1
          </div>
          <div>
            Count 0
          </div>
        </div>
        <div
          class="sc-eCstlR hYyfgm"
        >
          <div>
            Warning Type 2
          </div>
          <div>
            Count 0
          </div>
        </div>
        <div
          class="sc-eCstlR hYyfgm"
        >
          <div>
            Info Type 3
          </div>
          <div>
            Count 1
          </div>
          <div
            class="sc-bdfBQB kdxtmC"
            type="2"
          >
            <span
              class="sc-gsTEea hKnFKv"
            >
              Provident omnis voluptas modi vel enim.
            </span>
            <span
              class="sc-dlfnuX einvaF"
            >
              Clear
            </span>
          </div>
        </div>
      </div>
      <div
        class="sc-pGacB fTQQLE"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
